src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateCopies = [];
+  let stateCopy = { ...state };
+
+  for (const action of actions) {
+    let nextState;
+
+    switch (action.type) {
+      case 'addProperties':
+        nextState = { ...stateCopy, ...action.extraData };
+        break;
+      case 'removeProperties':
+        nextState = { ...stateCopy };
+
+        for (const key of action.keysToRemove) {
+          if (nextState.hasOwnProperty(key)) {
+            delete nextState[key];
```

Comment:
The check `nextState.hasOwnProperty(key)` is unnecessary since the delete operator will not throw an error if the property does not exist. You can safely remove this check and directly delete the key.